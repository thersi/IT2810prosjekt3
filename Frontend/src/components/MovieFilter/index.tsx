import { MouseEvent, useState } from "react";
import IconButton from "@mui/material/IconButton";
import MenuItem from "@mui/material/MenuItem";
import Menu from "@mui/material/Menu";
import ArrowDownwardIcon from "@mui/icons-material/ArrowDownward";
import ArrowUpwardIcon from "@mui/icons-material/ArrowUpward";
import CheckIcon from "@mui/icons-material/Check";
import FilterListIcon from "@mui/icons-material/FilterList";
import { Divider } from "@mui/material";
import { MovieFilterProps } from "../../Interfaces/Interfaces";

export default function MovieFilter(props: MovieFilterProps) {
  const [isAsc, setIsAsc] =
    useState<boolean>(
      true
    ); /* isAsc has value true if list is to be sorted in ascending order, false if descending */
  const [filter, setFilter] =
    useState<string>(
      "title"
    ); /* Initially filtered by title, sets filter to be chosen value. */
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);

  const handleMenu = (event: MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget); /* Decides if popup menu open or closed */
  };

  const handleClose = () => {
    setAnchorEl(null); /* Closes menu */
  };

  const handleSort = (e: MouseEvent<HTMLElement>, value: boolean) => {
    setIsAsc(value);
    props.handleSort(value); /* Sends value to parent */
  };
  const handleFilter = (e: MouseEvent<HTMLElement>, value: string) => {
    setFilter(value);
    props.handleFilter(value); /* Sends value to parent */
  };
  return (
    <div>
      <IconButton
        id="sortButton"
        size="large"
        edge="start"
        color="inherit"
        aria-label="open drawer"
        sx={{ mr: 2 }}
        onClick={handleMenu}
      >
        <FilterListIcon />
      </IconButton>
      <Menu
        id="menu-appbar"
        anchorEl={anchorEl}
        anchorOrigin={{
          vertical: "top",
          horizontal: "right",
        }}
        keepMounted
        transformOrigin={{
          vertical: "top",
          horizontal: "right",
        }}
        open={Boolean(anchorEl)}
        onClose={handleClose}
      >
        <MenuItem disabled>Sort by order: </MenuItem>
        {/* Disabled for visuals */}
        <MenuItem
          id="checkAsc"
          value="asc"
          onClick={(_) => handleSort(_, true)}
        >
          {isAsc && <CheckIcon id="ascCheck" />}{" "}
          {/* if ascending, checkIcon is
          shown */}
          <ArrowUpwardIcon /> Asc
        </MenuItem>
        <MenuItem
          id="checkDesc"
          value="desc"
          onClick={(_) => handleSort(_, false)}
        >
          {!isAsc && <CheckIcon />} {/* If descending, checkIcon is shown */}
          <ArrowDownwardIcon /> Desc
        </MenuItem>
        <Divider />
        <MenuItem disabled>Sort by type: </MenuItem>
        <MenuItem
          id="checkTitle"
          value="Title"
          onClick={(_) =>
            handleFilter(_, "title")
          } /* Calls on function handlefilter to set filter value to 'title' */
        >
          {filter === "title" && <CheckIcon />}
          Title
        </MenuItem>
        <MenuItem
          id="checkYear"
          value="Year"
          onClick={(_) =>
            handleFilter(_, "year")
          } /* Calls on function handlefilter to set filter value to 'year' */
        >
          {filter === "year" && <CheckIcon />}{" "}
          {/* The value chosen is the one that has a checkmark */}
          Year
        </MenuItem>
      </Menu>
    </div>
  );
}
