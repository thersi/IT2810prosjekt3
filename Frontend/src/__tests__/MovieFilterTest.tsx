import EnzymeToJson from "enzyme-to-json";
import { configure, mount, render, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import MovieFilter from "../components/MovieFilter";
import "@testing-library/jest-dom";
import MenuItem from "@mui/material/MenuItem";

configure({ adapter: new Adapter() });
const component = (
  <MovieFilter
    handleSort={() => {
      return 1;
    }}
    handleFilter={() => {
      return "title";
    }}
  />
);
describe("Movie Filter", () => {
  it("renders correctly", () => {
    render(component);
  });
  it("Should match snapshot", () => {
    const snapshotCheck = shallow(component);
    expect(EnzymeToJson(snapshotCheck)).toMatchSnapshot();
  });
});
/* it("Should render correct number of menu Items", () => {
    const shallowComp = mount(component);
    expect(shallowComp.find("IconButton").exists()).toBeTruthy(); */
/* expect(shallowComp.find("MenuItem")).toHaveLength(6); */

/*  it("Should show the chosen menu item", () => {
    const shallowComp = shallow(component);
    expect(shallowComp.find("#ascCheck").parent).toBe(
      shallowComp.find("#checkAsc")
    ); */

//descSort.simulate("click");

/* 
it("should contain correct components", () => {
  expect(component.find("SearchBar").exists()).toBeTruthy();
  expect(component.find("MovieFilter").exists()).toBeTruthy();
  expect(component.find("MovieDialog").exists()).toBeFalsy();
});
 */
