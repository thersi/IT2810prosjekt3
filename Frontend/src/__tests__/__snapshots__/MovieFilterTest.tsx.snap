// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Filter Should match snapshot 1`] = `
<div>
  <ForwardRef(IconButton)
    aria-label="open drawer"
    color="inherit"
    edge="start"
    id="sortButton"
    onClick={[Function]}
    size="large"
    sx={
      Object {
        "mr": 2,
      }
    }
  >
    <Memo(ForwardRef(FilterListIcon)) />
  </ForwardRef(IconButton)>
  <ForwardRef(Menu)
    anchorEl={null}
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    id="menu-appbar"
    keepMounted={true}
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
  >
    <ForwardRef(MenuItem)
      disabled={true}
    >
      Sort by order: 
    </ForwardRef(MenuItem)>
    <ForwardRef(MenuItem)
      id="checkAsc"
      onClick={[Function]}
      value="asc"
    >
      <Memo(ForwardRef(CheckIcon))
        id="ascCheck"
      />
      <Memo(ForwardRef(ArrowUpwardIcon)) />
       Asc
    </ForwardRef(MenuItem)>
    <ForwardRef(MenuItem)
      id="checkDesc"
      onClick={[Function]}
      value="desc"
    >
      <Memo(ForwardRef(ArrowDownwardIcon)) />
       Desc
    </ForwardRef(MenuItem)>
    <ForwardRef(Divider) />
    <ForwardRef(MenuItem)
      disabled={true}
    >
      Sort by type: 
    </ForwardRef(MenuItem)>
    <ForwardRef(MenuItem)
      id="checkTitle"
      onClick={[Function]}
    >
      Title
    </ForwardRef(MenuItem)>
    <ForwardRef(MenuItem)
      id="checkYear"
      onClick={[Function]}
    >
      Year
    </ForwardRef(MenuItem)>
  </ForwardRef(Menu)>
</div>
`;
